agent/src/capability/baseconfigstore.cpp

```diff
@@ -23,173 +23,201 @@ namespace softwarecontainer {
+void BaseConfigStore::parseCapabilities(json_t *capabilities)
 {
     size_t i;
     json_t *capability;
+    std::string errorMessage;
 
     log_debug() << "Size of capabilities is " << std::to_string(json_array_size(capabilities));
 
     json_array_foreach(capabilities, i, capability) {
         if (!json_is_object(capability)) {
-            log_error() << "Capability is not a json object";
-            return ReturnCode::FAILURE;
+            errorMessage = "Capability is not a json object";
+            log_error() << errorMessage;
+            throw CapabilityParseError("Capability parse error - " + errorMessage);
         }
 
         std::string capName;
         if (!JSONParser::read(capability, "name", capName)) {
-            log_error() << "Could not read Capability name";
-            return ReturnCode::FAILURE;
+            errorMessage = "Could not read Capability name";
+            log_error() << errorMessage;
+            throw CapabilityParseError("Capability parse error - " + errorMessage);
         }
 
         json_t *gateways = json_object_get(capability, "gateways");
         if (nullptr == gateways) {
-            log_error() << "Could not read Gateways";
-            return ReturnCode::FAILURE;
+            errorMessage = "Could not read Gateways";
+            log_error() << errorMessage;
+            throw CapabilityParseError("Capability parse error - " + errorMessage);
         }
         if (!json_is_array(gateways)) {
-            log_error() << "Gateways is not an array";
-            return ReturnCode::FAILURE;
+            errorMessage = "Gateways is not an array";
+            log_error() << errorMessage;
+            throw CapabilityParseError("Capability parse error - " + errorMessage);
         }
 
         log_debug() << "Found capability \"" << capName << "\", parsing gateways...";
 
-        ReturnCode parseGatewayResult = parseGatewayConfigs(capName, gateways);
-        if (isError(parseGatewayResult)) {
-            return ReturnCode::FAILURE;
+        // TODO (ethenor): Is this correct?
+        try {
+            parseGatewayConfigs(capName, gateways);
+        } catch(CapabilityParseError &error) {
+            throw;
         }
     }
-    return ReturnCode::SUCCESS;
+    //return;
 }
 
-ReturnCode BaseConfigStore::parseGatewayConfigs(std::string capName, json_t *gateways)
+void BaseConfigStore::parseGatewayConfigs(std::string capName, json_t *gateways)
 {
     if (m_capMap.count(capName) > 0) {
         log_debug() << "Capability " << capName << " already loaded.";
-        return ReturnCode::SUCCESS;
+        return;
     }
 
     size_t i;
     json_t *gateway;
+    std::string errorMessage;
 
     GatewayConfiguration gwConf;
     json_array_foreach(gateways, i, gateway) {
         if (!json_is_object(gateway)) {
-            log_error() << "Gateway is not a json object";
-            return ReturnCode::FAILURE;
+            errorMessage = "Gateway is not a json object";
```

Comment:
I think this message could be more to the point, e.g what has gone wrong. We should not use the word "Gateway" like this, since it's a very domain specific thing. In this case, we happen to have a key with the same word, but it's the "gateway" json key, not a "Gateway". I'd like to reduce the possibility of confusion here.